jmclient/jmclient/wallet_service.py

```diff
@@ -0,0 +1,362 @@
+        Before starting the event loop, we cache
+        the current most recent transactions as
+        reported by the blockchain interface, since
+        we are interested in deltas.
+        """
+        self.bci.sync_wallet(self.wallet, fast=fast,
+                             restart_cb=self.restart_callback)
+        self.synced = self.bci.wallet_synced
+        # Don't attempt updates on transactions that existed
+        # before startup
+        self.old_txs = self.bci.list_transactions(100)
+        return self.synced
+
+    def register_callbacks(self, callbacks, txinfo, cb_type="all"):
+        """ Register callbacks that will be called by the
+        transaction monitor loop, on transactions stored under
+        our wallet label (see WalletService.get_wallet_name()).
+        Callback arguments are currently (txd, txid) and return
+        is boolean, except "confirmed" callbacks which have
+        arguments (txd, txid, confirmations).
+        Note that callbacks MUST correctly return True if they
+        recognized the transaction and processed it, and False
+        if not. The True return value will be used to remove
+        the callback from the list.
+        Arguments:
+        `callbacks` - a list of functions with signature as above
+        and return type boolean.
+        `txinfo` - either a txid expected for the transaction, if
+        known, or a tuple of the ordered output set, of the form
+        (('script': script), ('value': value), ..). This can be
+        constructed from jmbitcoin.deserialize output, key "outs",
+        using tuple(). See WalletService.transaction_monitor().
+        `cb_type` - must be one of "all", "unconfirmed", "confirmed";
+        the first type will be called back once for every new
+        transaction, the second only once when the number of
+        confirmations is 0, and the third only once when the number
+        of confirmations is > 0.
+        """
+        if cb_type == "all":
+            # note that in this case, txid is ignored.
+            self.callbacks["all"].extend(callbacks)
+        elif cb_type in ["unconfirmed", "confirmed"]:
+            if txinfo not in self.callbacks[cb_type].keys():
+                self.callbacks[cb_type][txinfo] = []
+            self.callbacks[cb_type][txinfo].extend(callbacks)
+        else:
+            assert False, "Invalid argument: " + cb_type
+
+
+    def start_wallet_monitoring(self, syncresult):
+        """ Once the initialization of the service
+        (currently, means: wallet sync) is complete,
+        we start the main monitoring jobs of the
+        wallet service (currently, means: monitoring
+        all new transactions on the blockchain that
+        are recognised as belonging to the Bitcoin
+        Core wallet with the JM wallet's label).
+        """
+        if not syncresult:
+            jlog.error("Failed to sync the bitcoin wallet. Shutting down.")
+            reactor.stop()
```

Comment:
Does `reactor.stop()` immediately stop execution of the whole program? Otherwise, shouldn't there be a `return`?

Reply:
Good catch. Fixed in https://github.com/JoinMarket-Org/joinmarket-clientserver/pull/359/commits/7637865a7f987e40f80a22209a5efd40ba6f34b5